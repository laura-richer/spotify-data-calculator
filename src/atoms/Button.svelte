<script lang="ts">
  export let btnText: string = 'Add button text';
  export let btnType: string = 'primary';
  export let btnHoverColor: string = 'green';
</script>

<button on:click class="btn btn--{btnType} btn-focus--{btnHoverColor}">{btnText}</button>
