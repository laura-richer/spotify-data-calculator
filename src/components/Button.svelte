<script lang="ts">
	export let active = false;
	export let buttonText = 'Add button text';
	export let buttonType = 'primary';
	export let buttonHoverColor = 'green';
</script>

<button on:click class="btn btn--{buttonType} btn-focus--{buttonHoverColor}" class:btn--active={active}
	>{buttonText}</button
>

<style lang="scss">
@use 'sass:math';
@import '../scss/resources';

.btn {
  $btn: &;
  @include border-outset;


  position: relative;
  background-color: var(--color-grey);
  cursor: pointer;

  &:hover,
  &:focus,
  &--active {
    @include border-inset;
  }

  &--active {
    &::after {
      position: absolute;
      top: 1rem;
      right: 1rem;
      width: 11px;
      height: 15px;
      background: url(../assets/images/flag.png) center center no-repeat;
      content: '';
    }
  }

  &--primary {
    padding: 0.8rem $spacer;
    border-width: 5px;
    font-size: 1.2rem;
  }

  &--compact {
    padding: 0.7rem $spacer;
    border-width: 3px;
    font-size: 0.8rem;

    &::after {
      top: $spacer * 0.5;
    }
  }

  // TODO refactor to use css variables like blueprint
  &-focus--green {
    &:focus,
    &#{$btn}--active {
      color: var(--color-green);
    }
  }

  &-focus--blue {
    &:focus,
    &#{$btn}--active {
      color: var(--color-blue);
    }
  }

  &-focus--red {
    &:focus,
    &#{$btn}--active {
      color: var(--color-red);
    }
  }
}
</style>
