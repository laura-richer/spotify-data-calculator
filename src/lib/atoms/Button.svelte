<script lang="ts">
	export let buttonSelected = false;
  export let buttonDisabled = false;
	export let buttonText = 'Add button text';
	export let buttonType = 'primary';
  export let buttonHoverColor: string;
</script>

<button
  on:click
  class="btn btn--{buttonType} btn-hover--{buttonHoverColor}"
  class:btn--active={buttonSelected}
  class:btn--disabled={buttonDisabled}
>
  {buttonText}
</button>

<style lang="scss">
@use 'sass:math';
@import '../scss/resources';

.btn {
  $btn: &;
  @include border-outset;


  position: relative;
  background-color: var(--color-grey);
  cursor: pointer;

  &:hover,
  &:focus,
  &--active {
    @include border-inset;
  }

  &--active {
    &::after {
      position: absolute;
      top: 1rem;
      right: 1rem;
      width: 11px;
      height: 15px;
      background: url(../assets/images/flag.png) center center no-repeat;
      content: '';
    }
  }

  &--primary {
    padding: 0.8rem $spacer;
    border-width: 5px;
    font-size: 1.2rem;
  }

  &--compact {
    padding: 0.7rem $spacer;
    border-width: 3px;
    font-size: 0.8rem;

    &::after {
      top: $spacer * 0.5;
    }
  }

  &-hover--green {
    &:hover,
    &#{$btn}--active {
      color: var(--color-green);
    }
  }

  &-hover--blue {
    &:hover,
    &#{$btn}--active {
      color: var(--color-blue);
    }
  }

  &-hover--red {
    &:hover,
    &#{$btn}--active {
      color: var(--color-red);
    }
  }

  &--disabled {
    @include border-inset;

    cursor: pointer;
    opacity: 0.2;
    pointer-events: none;
  }
}
</style>
