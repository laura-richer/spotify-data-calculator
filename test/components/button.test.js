import '@testing-library/jest-dom';
import { render, screen, fireEvent } from "@testing-library/svelte";
import Button from '../../src/components/Button.svelte';

const buttonText = 'Test button';
const buttonActiveColor = 'red';

describe('Button component', function () {
  it('renders to match snapshot', () => {
    const result = Button;
    expect(result).toMatchSnapshot()
  });

  it('renders in the DOM', () => {
    render(Button);
    expect(Button).toBeTruthy();
  });

  // it('should be a button element', () => {
  //   render(Button, { buttonText });
  //   const button = screen.getByText(buttonText);
  //   expect(button.tag).toEqual('button');
  // });

  it('should render with primary class by default', () => {
    render(Button, { buttonText });
    const button = screen.getByText(buttonText);

    expect(button).toHaveClass('btn');
    expect(button).toHaveClass('btn--primary');
  });

  it('should render with green active state by default', () => {
    render(Button, { buttonText });
    const button = screen.getByText(buttonText);

    expect(button).toHaveClass('btn');
    expect(button).toHaveClass('btn-focus--green');
  });

  it('should not have active class by default', () => {
    render(Button, { buttonText });
    const button = screen.getByText(buttonText);

    expect(button).not.toHaveClass('btn--active');
  });

  it('should render correct classes with props', () => {
    render(Button, { buttonText, buttonActiveColor, buttonType: 'compact' });
    const button = screen.getByText(buttonText);

    expect(button).toHaveClass('btn');
    expect(button).toHaveClass('btn--compact');
    expect(button).toHaveClass(`btn-focus--${buttonActiveColor}`);
    expect(button.innerHTML).toBe(buttonText);
  });

  it('should render active class when active prop is passed', () => {
    render(Button, { buttonText, active: true });
    const button = screen.getByText(buttonText);

    expect(button).toHaveClass('btn');
    expect(button).toHaveClass('btn--active');
  });

  // it('should have a colour variable to match the active prop color', () => {
  //   render(Button, { buttonText, buttonActiveColor, active: true });
  //   const button = screen.getByText(buttonText);

  //   expect(button).toHaveClass(`btn-focus--${buttonActiveColor}`);
  //   expect(button).toHaveStyle(`color: var(--color-${buttonActiveColor})`);
  // });

  // it('should emit click when clicked', async () => {
  //   render(Button, { buttonText });
  //   const button = screen.getByText(buttonText);

  //   await fireEvent.click(button);

  // });
});



