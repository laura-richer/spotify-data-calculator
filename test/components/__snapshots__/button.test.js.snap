// Vitest Snapshot v1

exports[`Button component > renders to match snapshot 1`] = `[Function]`;

exports[`Counter component > renders to match snapshot 1`] = `[Function]`;

exports[`renders to match snapshot 1`] = `[Function]`;
